src/workerd/api/worker-loader.c++

```diff
@@ -0,0 +1,219 @@
+#include <workerd/io/compatibility-date.h>
+#include <workerd/io/features.h>
+#include <workerd/io/io-context.h>
+
+#include <capnp/message.h>
+
+namespace workerd::api {
+
+class WorkerStubEntrypointOutgoingFactory final: public Fetcher::OutgoingFactory {
+ public:
+  WorkerStubEntrypointOutgoingFactory(kj::Own<IoChannelFactory::SubrequestChannel> channel)
+      : channel(kj::mv(channel)) {}
+
+  kj::Own<WorkerInterface> newSingleUseClient(kj::Maybe<kj::String> cfStr) override {
+    auto& context = IoContext::current();
+
+    return context.getMetrics().wrapSubrequestClient(context.getSubrequest(
+        [&](TraceContext& tracing, IoChannelFactory& ioChannelFactory) {
+      return channel->startRequest({.cfBlobJson = kj::mv(cfStr), .tracing = tracing});
+    },
+        {.inHouse = true,
+          .wrapMetrics = true,
+          .operationName = kj::ConstString("dynamic_worker_subrequest"_kjc)}));
+  }
+
+ private:
+  kj::Own<IoChannelFactory::SubrequestChannel> channel;
+};
+
+jsg::Ref<Fetcher> WorkerStub::getEntrypoint(jsg::Lock& js,
+    jsg::Optional<kj::Maybe<kj::String>> name,
+    jsg::Optional<EntrypointOptions> options) {
+  Frankenvalue props;
+
+  KJ_IF_SOME(o, options) {
+    KJ_IF_SOME(p, o.props) {
+      props = Frankenvalue::fromJs(js, p);
+    }
+  }
+
+  kj::Maybe<kj::String> entrypointName;
+  KJ_IF_SOME(n, name) {
+    KJ_IF_SOME(n2, n) {
+      if (n2 != "default"_kj) {
+        entrypointName = kj::mv(n2);
+      }
+    }
+  }
+
+  kj::Own<Fetcher::OutgoingFactory> factory = kj::heap<WorkerStubEntrypointOutgoingFactory>(
+      channel->getEntrypoint(kj::mv(entrypointName), kj::mv(props)));
+
+  return js.alloc<Fetcher>(
+      IoContext::current().addObject(kj::mv(factory)), Fetcher::RequiresHostAndProtocol::YES, true);
+}
+
+jsg::Ref<DurableObjectClass> WorkerStub::getDurableObjectClass(jsg::Lock& js,
+    jsg::Optional<kj::Maybe<kj::String>> name,
+    jsg::Optional<EntrypointOptions> options) {
+  Frankenvalue props;
+
+  KJ_IF_SOME(o, options) {
+    KJ_IF_SOME(p, o.props) {
+      props = Frankenvalue::fromJs(js, p);
+    }
+  }
+
+  kj::Maybe<kj::String> entrypointName;
+  KJ_IF_SOME(n, name) {
+    KJ_IF_SOME(n2, n) {
+      if (n2 != "default"_kj) {
+        entrypointName = kj::mv(n2);
+      }
+    }
+  }
+
+  return js.alloc<DurableObjectClass>(IoContext::current().addObject(
+      channel->getActorClass(kj::mv(entrypointName), kj::mv(props))));
+}
+
+class NullGlobalOutboundChannel: public IoChannelFactory::SubrequestChannel {
+ public:
+  kj::Own<WorkerInterface> startRequest(IoChannelFactory::SubrequestMetadata metadata) override {
+    JSG_FAIL_REQUIRE(Error,
+        "This worker is not permitted to access the internet via global functions like fetch(). "
+        "It must use capabilities (such as bindings in 'env') to talk to the outside world.");
+  }
+};
+
+jsg::Ref<WorkerStub> WorkerLoader::get(
+    jsg::Lock& js, kj::String name, jsg::Function<jsg::Promise<WorkerCode>()> getCode) {
```

Comment:
What about providing the `name`/`id` to the `getCode` callback? That'd make it easier to dynamically load a worker from external storage.

Reply:
You're saying instead of capturing the name into the callback like:

```
env.loader.get(name, () => load(name));
```

You want to be able to write:

```
env.loader.get(name, load);
```

Is it really a big deal, though? I think it would actually perform worse: converting the name back from C++ to JavaScript would be slower than creating the lambda that captures it. (Though the difference is negligible regardless.)